url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/file/FORMFile.java

```
@@ -0,0 +1,203 @@
+		//only FORM chunks can be added at root level
+		if (!name.equals("FORM")) {
+			throw new IOException(KlangConstants.ERROR_NOT_FORM);
+		}
+		if (position < 1) {
+			position = 1;
+		}
+		if (position > this.getChunks().size() + 1) {
+			position = this.getChunks().size() + 1;
+		}
+		long start = -1;
+		if (position == 1) {
+			start = 0;
+		} else {
+			start = this.getChunks().get(position - 2).getEndPosition();
+		}
+		//first, add header to a tempfile, then add bytes and possibly a pad byte
+		File tempFile = File.createTempFile("klang", ".data");
+		RandomAccessFile tempraf = new RandomAccessFile(tempFile, "rw");
+		try {
+			tempraf.seek(0);
+			//header
+			FourCC fccname = new FourCC();
+			fccname.setValueFromString(name);
+			fccname.writeValueToFile(tempraf);
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			Long sizeval = KlangUtil.adjustForPadByte(bytes.length);
+			size.setValue(sizeval);
+			size.writeValueToFile(tempraf);
+		} catch (BadValueException err) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		//now write bytes to temp file
+		KlangUtil.appendToFile(bytes, tempFile);
+		if (KlangUtil.adjustForPadByte(bytes.length) != bytes.length) {
+			//add pad byte
+			KlangUtil.appendToFile(new byte[1], tempFile);
+		}
+		//now merge tempfile into file
+		KlangUtil.insertIntoFile(tempFile, this.getFile(), start);
+		tempFile.delete();
+		reparseFile(raf);		
+	}
+	
+	public void chunkWillChangeSize(Chunk chunk, long newsize, RandomAccessFile raf) throws IOException {
+		//no need to do anything here
+	}
+
+	public void chunkHasChanged(Chunk chunk, RandomAccessFile raf) throws IOException {
+		//no need to do anything here
+	}
+	
+	public void editChunkName(Chunk chunk, String newName, RandomAccessFile raf) throws IOException {
+		try {
+			raf.seek(chunk.getStartPosition());
+			FourCC fcc = new FourCC();
+			fcc.setValueFromString(newName);
+			fcc.writeValueToFile(raf);
+		} catch (BadValueException err) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)